### PR TITLE
Feature/customfield semester

### PR DESCRIPTION
**Pull request by NinaHerrmann**
## **WIP**
This pull request is WIP because I don't want it to be submitted on Friday afternoon (sleep over it and submit it on Monday).

**Feature finished:**

- Adds a dropdown in case the customfield 'Semester' exists as a select option. The customfield is expected to have at key 1 the default semester (which is most probably no semester). 

- Removes deprecated `pix_url` function

**Features for the future:**

- Extract the suitable semester from the request to the database. 

**Additional information:**
A customfield which defines the semester can be desirable for multiple purposes. One of the use cases is the myoverview block of the moodle core where courses can be filtered according to customfields. Another use case might be navigation to structure the courses according to their semesters. 
For any of those use cases, it is desirable to have the correct value in the customfield which requires to allow teachers to set this value when they are requesting a course via the plugin. This is the feature included in this request. 
The implementation might change in the future since an own customfieldtype might be more suitable to improve the usability.

### CODE REVIEW

File: handle_customfield_semester.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    global $DB;
+    $customfield = $DB->get_record('customfield_field', array('name' =>  'Semester', 'type' => 'select'));
+    $customfieldcontroller = \customfield_date\field_controller::create($customfield->id);
+    $configdata = $customfieldcontroller->get('configdata');
+    $semesterinarray = explode("\n", $configdata['options']);
+    if (array_key_exists($currentsemester, $semesterinarray)){
+        $previouscustomfield = $DB->get_record('customfield_data', array('instanceid' => $courseid));
+        $numericalrepresentation = $currentsemester + 1;
+        // In case we have data for a previous field update in case it changed.
+        if ($DB->get_record('customfield_data', array('instanceid' => $courseid))) {
```
Comment (Dagefoerde) [inline handle_customfield_semester.php]: Shouldn't the WHERE contain the name of the field / field id as well?

File: handle_customfield_semester.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            if (!$previouscustomfield->value == $numericalrepresentation) {
+                $previouscustomfield->value = $numericalrepresentation;
+                $DB->update_record('customfield_data', $previouscustomfield);
+            }
+        } else {
+            // Otherwise create an object and insert it into table.
+            $context = $DB->get_record('context', array('contextlevel' => CONTEXT_COURSE, 'instanceid' => $courseid));
```
Comment (Dagefoerde) [inline handle_customfield_semester.php]: `context_course::instance($courseid)` is better

File: handle_customfield_semester.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                'valueformat' => 0,
+                'timecreated' => $currenttimestamp,
+                'timemodified' => $currenttimestamp,
+                'contextid' => $context->id
+            ];
+            $DB->insert_record('customfield_data', $dataobject);
+        }
+    } else {
+        // The key does not exist in the available choices so we will make it a no semester course.
+        $context = $DB->get_record('context', array('contextlevel' => CONTEXT_COURSE, 'instanceid' => $courseid));
```
Comment (Dagefoerde) [inline handle_customfield_semester.php]: see above

File: handle_customfield_semester.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        $currenttimestamp = time();
+        $dataobject = (object) [
+            'fieldid' => $customfield->id,
+            'instanceid' => $courseid,
+            'intvalue' => 1,
```
Comment (Dagefoerde) [inline handle_customfield_semester.php]: is this a magic constant for semester-independent courses? I'd prefer making this explicit by creating a variable

File: lang/de/local_lsf_unification.php [hunk 5]
```diff
@@ -88,6 +88,8 @@
 $string['config_category'] = "Kategorie";
 $string['config_category_wish'] = "Kategorie-Umzugs-Wunsch";
 $string['config_category_wish_help'] = "Falls Sie den Kurs gerne in einer anderen st&auml;rker spezifizierten Kategorie eingeordnet haben m&ouml;chten, hinterlassen Sie bitte hier einen Kommentar  mit der entsprechenden Wunschkategorie und -pfad.";
+$string['config_course_semester'] = "Semester";
+$string['config_course_semester_missing'] = "Die Feld zur Angabe des Semesters ist nicht ausgef&uuml;llt.";
```
Comment (Dagefoerde) [inline lang/de/local_lsf_unification.php]: *Das Feld

File: lang/en/local_lsf_unification.php [hunk 6]
```diff
@@ -89,6 +89,8 @@
 $string['config_category'] = "Category";
 $string['config_category_wish'] = "Category Relocation Wish";
 $string['config_category_wish_help'] = "If you have a wish to get your course moved into a more specific category, please leave a comment here containing your wish-category and path.";
+$string['config_course_semester'] = "Semester";
+$string['config_course_semester_missing'] = "The field defining the current semester is missing.";
```
Comment (Dagefoerde) [inline lang/en/local_lsf_unification.php]: *Semester -> Term
*the current semester -> the term that the course belongs to

File: handle_customfield_semester.php (hunk 2 above)
Comment (NinaHerrmann) [reply handle_customfield_semester.php]: I think the best way to keep this documented is to link fixes 
https://github.com/NinaHerrmann/his_unification/blob/5d18a0fe3f1b46caeae24a3908db44e9c2e85fef/handle_customfield_semester.php#L39

File: handle_customfield_semester.php (hunk 3 above)
Comment (NinaHerrmann) [reply handle_customfield_semester.php]: as above: 
https://github.com/NinaHerrmann/his_unification/blob/5d18a0fe3f1b46caeae24a3908db44e9c2e85fef/handle_customfield_semester.php#L39

File: handle_customfield_semester.php (hunk 1 above)
Comment (NinaHerrmann) [reply handle_customfield_semester.php]: Yes also removed the doubled request:
https://github.com/NinaHerrmann/his_unification/blob/5d18a0fe3f1b46caeae24a3908db44e9c2e85fef/handle_customfield_semester.php#L42

File: handle_customfield_semester.php (hunk 4 above)
Comment (NinaHerrmann) [reply handle_customfield_semester.php]: https://github.com/NinaHerrmann/his_unification/blob/5d18a0fe3f1b46caeae24a3908db44e9c2e85fef/handle_customfield_semester.php#L67

File: lang/en/local_lsf_unification.php (hunk 6 above)
Comment (NinaHerrmann) [reply lang/en/local_lsf_unification.php]: https://github.com/NinaHerrmann/his_unification/blob/5d18a0fe3f1b46caeae24a3908db44e9c2e85fef/lang/en/local_lsf_unification.php#L92

**Comment by Dagefoerde** [comment]
:shipit:

**Pull request merged**